### DISCUSSION TITLE
MF6 SFR perioddata Input Format?

### DISCUSSION THREAD

**Discussion by jklakovich-INTERA** [modflowpy/flopy#1718] [Q&A]
Hello, 

I tried to input what I though to be "stress period data" to the perioddata parameter of flopy.mf6.ModflowGwfsfr - that is, a dict of lists of formatted strings. Here is the python input and output for perioddata (just one stress period here):

Input:
![image](https://user-images.githubusercontent.com/112116923/218810518-41de3ee4-3fba-4c98-b314-618e010fde0a.png)
Output:
![image](https://user-images.githubusercontent.com/112116923/218810710-1763e3e9-b1e4-4ede-8058-2bcbc96654c5.png)

Here is the error that I get:
![image](https://user-images.githubusercontent.com/112116923/218811487-ba2f8674-754c-47fa-893e-1368ab147f0b.png)

My format is wrong but I am having trouble deciphering from the documentation
(https://flopy.readthedocs.io/en/3.3.5/source/flopy.mf6.modflow.mfgwfsfr.html?highlight=sfr)
what is the correct input format. Could you please advise?

Thank you,
Jeff

**Comment by jklakovich-INTERA** [comment]
I also tried a dict of lists of tuples comprising an integer and a string - e.g., {0: [(102, 'status active'), ... ]} - and got the same error.

Thanks,
Jeff

**Comment by christianlangevin** [comment]
How about something like this:

```
    perioddata = [
        [0, "status", "active"],
        [1, "status", "active"],
        [2, "status", "active"],
        [3, "status", "active"],
        [4, "status", "active"],
        [0, "diversion", 0, divflow],
        [1, "diversion", 0, divflow],
        [2, "diversion", 0, divflow2],
        [3, "diversion", 0, divflow3],
    ]
```

**Comment by jklakovich-INTERA** [comment]
Hi @langevin-usgs,

I input something similar -

![image](https://user-images.githubusercontent.com/112116923/218820533-f29a5196-1357-4b6d-8c63-f34b5938cb19.png)

And got the same error - 
![image](https://user-images.githubusercontent.com/112116923/218820633-e37c6b52-c609-4486-9640-f171331b4fd2.png)

Also, perioddata must index the stress period, I think, so I'm not sure how a list of lists could have worked.

Jeff

**Comment by jklakovich-INTERA** [comment]
@langevin-usgs, I tried a variant of that - a dict of lists of lists:

Input:
![image](https://user-images.githubusercontent.com/112116923/218821997-7e09eb94-b680-46ca-bb47-1f0775e4cb0b.png)

Output:
![image](https://user-images.githubusercontent.com/112116923/218822173-5fe14638-13a3-4680-8a9d-0a9917cd9ba4.png)

Still got the same error....

**Comment by christianlangevin** [comment]
There are a bunch of examples in this folder here: https://github.com/MODFLOW-USGS/modflow6/tree/develop/autotest

I was looking at the `test_gwf_mvr01.py` test, which has the following functioning code.  Flopy should take a list and assume it is for stress period 0, or you can provide a dictionary.

![image](https://user-images.githubusercontent.com/3372403/218823070-77e1a57c-f95f-4a61-b480-2dc73f85cc26.png)

**Reply by jklakovich-INTERA** [reply]
Hi @langevin-usgs, I tried to input the variable 'p_data' from autotest/test_gwf_buy_sfr01.py, shown below, into my perioddata variable:

![image](https://user-images.githubusercontent.com/112116923/218828503-ec50e026-116e-4cad-8fd7-d8379d8d5e4d.png)

... still get same error.

**Comment by jlarsen-usgs** [comment]
@jklakovich-INTERA 

Here's what I've done in before to create perioddata for SFR.
```python
    period_data = {
        i: [(0, "INFLOW", 50),] for i in range(12)
    }
```
Maybe the record needs to be in a tuple, even though FloPy should be able to handle any sort of iterable as input in these cases

**Comment by jklakovich-INTERA** [comment]
Hi @jlarsen-usgs,

I copied your code exactly - 

![image](https://user-images.githubusercontent.com/112116923/218825954-746f05f1-88f8-479f-a6df-fc423898c357.png)

And got the same error - 

![image](https://user-images.githubusercontent.com/112116923/218826056-be39b4f9-1a83-4f33-99a9-01ee5ade8caf.png)

Maybe it's a bug?

I am using 
FloPy 3.3.5
Microsoft Windows 10 Pro 10.0.19044 Build 19044

Strangely, I am not able to update FloPy using "conda update flopy" because I may not have the 'conda infrastructure' in my virtual environment (StackOverflow). I just discovered this while looking up my conda version.

Thanks for your help with this,

Jeff

**Reply by jlarsen-usgs** [reply]
@jklakovich-INTERA 
Interesting, I haven't seen that before. Maybe try upgrading your flopy to the latest development version (3.3.7) using pip.

```pip install -U https://github.com/modflowpy/flopy/zipball/develop```

**Comment by jklakovich-INTERA** [comment]
Flopy 3.3.7 is handling the SFR perioddata with no issues.

Thanks for your help @jlarsen-usgs, @langevin-usgs .

Jeff
